Install manpages via make install, also add some autogenerated manpages

nomnombtc opened the pull request:
This adds a new option `--enable-man` to configure, so manpages can be installed via `make install`.
This option defaults to yes, but opt out of the manpage installation is possible by running `./configure --disable-man` (or `--enable-man=no`) 

As the license situation of the contrib/debian manpages is that most of them are GPL it was not 100% clear if they can be included here, also they are incomplete anyway so I let `help2man` create a few manpages from the 0.13.0 binaries and included them.

The script used to generate them has also been added to contrib/devtools.

This targets #7626 but was also briefly discussed in #8568

Commit:
add script to generate manpages with help2man

Commit:
add gen-manpages.sh description to README.md

Commit:
add autogenerated manpages by help2man

Commit:
add doc/man/Makefile.am to include manpages

Commit:
add doc/man to subdir if configure flag --enable-man is set

Commit:
add conditional for --enable-man, default is yes

luke-jr commented on configure.ac:
```diff
@@ -167,6 +167,13 @@ AC_ARG_ENABLE([zmq],
 
+# Enable manpages
+AC_ARG_ENABLE(man,
+    [AS_HELP_STRING([--enable-man],
```
Normally we use --disable-* for the help for options which are enabled by default.

nomnombtc replied:
Makes sense after looking at the other configure options. Would this be better:
```
+    [AS_HELP_STRING([--disable-man],
+                    [do not install man pages (default is to install)])],,
```
And should I remove "# Enable manpages" comment? only --disable-wallet and --debug has a comment everything else doesn't.

luke-jr replied:
Sounds good to me

luke-jr commented on configure.ac:
```diff
@@ -167,6 +167,13 @@ AC_ARG_ENABLE([zmq],
+AC_ARG_ENABLE(man,
+    [AS_HELP_STRING([--enable-man],
+                    [install man pages [default=yes]])],,
+    enable_man=yes)
```
(I wonder if it should perhaps be disabled by default when building for Windows?)

luke-jr commented on contrib/devtools/README.md:
```diff
@@ -40,6 +40,12 @@ would be changed to:
+===============
+
+A small script to automatically create manpages in ../../doc/man by running the release binaries with the -help option.
+This requires help2man which can be found at: https://www.gnu.org/software/help2man/
```
Should probably mention the relevant binaries need to be in PATH, adjust it so the BITCOIND and/or BITCOINCLI environment variables are used, and/or just use the binaries in the build dir.

nomnombtc replied:
Hmm, the script was already a bit ugly, but it was very short which was the good thing about it :)
Would you like something like this better: http://pastebin.com/j6JFaiaq

Or I could leave it like it is and write:
>A small script to automatically create manpages in ../../doc/man by running binaries in $PATH with the --help option.

theuni replied:
Why not use a makefile target instead? All of that info would be readily available there.

nomnombtc replied:
> Why not use a makefile target instead? All of that info would be readily available there.

Do you mean something like [this](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/split-debug.sh.in) where it fills in some stuff in the script, or do you mean integrating help2man into the make buildprocess?

Because you can do this: https://www.gnu.org/software/help2man/#Makefile-usage 
until 5 minutes ago I assumed everyone building from source then needs help2man installed,
but this sentence suggest otherwise:
>This usage allows a manual page to be generated by the maintainer and included in the distribution without requiring the end-user to have help2man installed.

Now I am a bit confused :) Guess this needs more research.

nomnombtc replied:
@theuni 
I am playing around right now with the `help2man` Makefile method and it seems it indeed only
runs `help2man`:  *if* the manpage files don't exist yet **OR** a string in the source cpp file changed (I have no idea how it does this - blackmagic?).

So if the man pages exist and the help options matches the source files, the end user can build  without having `help2man` installed. If this confirms it seems much better than the shell script... 

But it has some side effects, now the commit id is back in the manpage. And some other small problem which my script was working around (matching copyright footer for all manpages)

theuni replied:
@nomnombtc I think they're just saying that if you commit pre-generated files to the repo, users don't have to do it.

For a makefile target, i mean something like this: https://github.com/bitcoin/bitcoin/blob/master/src/Makefile.qt.include#L408

In this case, it would look something like (in src/Makefile.am):
```Makefile
update-man: $(bin_PROGRAMS)
       $(AM_V_at)echo "[COPYRIGHT]" > footer.h2m
       $(AM_V_at)echo "Copyright $(COPYRIGHT_YEAR) $(COPYRIGHT_HOLDERS_FINAL)" >> footer.h2m
       $(AM_V_at)for prog in $^; do help2man -N --version-string=$(PACKAGE_VERSION) --include=footer.h2m -o $(top_builddir)/doc/man/`basename $$prog`.1 $(builddir)/$$prog$(EXEEXT); done
```
(that's not really tested)

Since it's in a makefile, you have all of the vars at your disposal without trying to guess program names/paths.

nomnombtc replied:
@theuni 

Oh ok now I get it, I was playing around with something like this:

```
bitcoind.1: $(top_builddir)/src/bitcoind.cpp
        -$(HELP2MAN) --no-info --version-string=$(PACKAGE_VERSION) \
          --output=$@ --name='manual page for $(BITCOIN_DAEMON_NAME) $(PACKAGE_VERSION)' \
          $(top_builddir)/src/$(BITCOIN_DAEMON_NAME)$(EXEEXT)
        -sed -i 's/\-$(shell git rev-parse HEAD | cut -c1-7)//g' $@

bitcoin-cli.1: $(top_builddir)/src/bitcoin-cli.cpp
        -$(HELP2MAN) --no-info --version-string=$(PACKAGE_VERSION) \
          --output=$@ --name='manual page for $(BITCOIN_CLI_NAME) $(PACKAGE_VERSION)' \
          $(top_builddir)/src/$(BITCOIN_CLI_NAME)$(EXEEXT)
        -sed -i 's/\-$(shell git rev-parse HEAD | cut -c1-7)//g' $@

bitcoin-tx.1: $(top_builddir)/src/bitcoin-tx.cpp
        -$(HELP2MAN) --no-info --version-string=$(PACKAGE_VERSION) \
          --output=$@ --name='manual page for $(BITCOIN_TX_NAME) $(PACKAGE_VERSION)' \
          $(top_builddir)/src/$(BITCOIN_TX_NAME)$(EXEEXT)
        -sed -i 's/\-$(shell git rev-parse HEAD | cut -c1-7)//g' $@

dist_man1_MANS=bitcoind.1 bitcoin-cli.1 bitcoin-tx.1

if ENABLE_QT
bitcoin-qt.1: $(top_builddir)/src/qt/bitcoin.cpp
        -$(HELP2MAN) --no-info --version-string=$(PACKAGE_VERSION) \
          --output=$@ --name='manual page for $(BITCOIN_GUI_NAME) $(PACKAGE_VERSION)' \
          $(top_builddir)/src/qt/$(BITCOIN_GUI_NAME)$(EXEEXT)
        -sed -i 's/\-$(shell git rev-parse HEAD | cut -c1-7)//g' $@
dist_man1_MANS+=bitcoin-qt.1
endif
```

Because my hope was to automate it completly, so that maybe the gitian build process generates the manpages in the end.It works pretty ok, but a huge problem is that running help2man on bitcoin-qt needs access to a running X-Server -_-.

So someone needs to generate and commit the manpages anyway. May as well than use something like your maketarget, at least this is much more elegant than my shell script... :)

Going to look into it, thanks!

luke-jr commented:
Why include generated files (the manpages in this case) in the git repo, rather than just building them during make? (configure can disable man by default if help2man isn't found)

laanwj commented:
> Why include generated files (the manpages in this case) in the git repo, rather than just building them during make? (configure can disable man by default if help2man isn't found)

Because this (now manual) step requires actually executing the built executables to get their help output. I disagree with making that a default step in the build process. It will also not work with cross-compilation.

laanwj commented:
utACK, will test.

MarcoFalke commented:
Agree with @laanwj that ​it is helpful to differentiate between generation of the man pages and installation. Keep in mind that this allows to fiddle with the output, in case help2man is not perfect.

Commit:
change help string --enable-man to --disable-man

jonasschnelli commented:
Concept ACK

laanwj commented:
Nits: 
- can we avoid the commit id being included in the generated manpages? This will generate commit noise every time they're regenerated for no good reason.
```       Bitcoin Core RPC client version v0.13.0.0-ga402396```
- `doc/release-process.md` should contain instructions to generate these pages (after bumping the version!)
- `bitcoin-tx` has no man page.

Apart from that, tested ACK.

theuni commented:
Concept ACK.

Agreed with the points above. Though, if we're pre-generating the pages and keeping them in-repo, we need to be careful about \#ifdefs in help output.

For example: https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L317

Also, I'd much rather use a makefile target for generating, similar to the way the other generated files are produced. That helps to keep everything in sync since the vars are shared around.

laanwj commented:
> Though, if we're pre-generating the pages and keeping them in-repo, we need to be careful about #ifdefs in help output.

I just don't like the alternative. If this could be parsed from the C source *without* executing, I'd be fine with doing generation during build process.
In any case that can be done in the future, this is strictly an improvement to how things currently are.

Commit:
regenerated all manpages with commit tag stripped, also add bitcoin-tx

Commit:
improved gen-manpages.sh, includes bitcoin-tx and strips commit tag, now also runs binaries from build dir by default, added variables for more control

nomnombtc commented:
>can we avoid the commit id being included in the generated manpages? This will generate diff noise every time they're regenerated for no good reason. Bitcoin Core RPC client version v0.13.0.0-ga402396

Yes I have updated the script somewhat to strip the commit id from the help2man output. It now also has some variables like BITCOIND BITCOINCLI and so on, as luke-jr suggested above somewhere and it now runs the binaries from the build dir,  not from $PATH anymore.

>doc/release-process.md should contain instructions to generate these pages (after bumping the version!)

Okay. Should I just write that it needs to be regenerated prior to release like this?

```
[...]
Before every minor and major release:

Update bips.md to account for changes since the last release.
Update version in sources (see below)
Write release notes (see below)
Regenerate manpages with script in contrib/devtools
[...]
```

>bitcoin-tx has no man page.

Oops. Thanks for noticing this, somehow I forgot about bitcoin-tx. It is included now :)

I also noticed another annoyance:
When build with `--with-gui=no` it still installs the bitcoin-qt manpage...
If we want to prevent this, the Makefile.am in doc/man needs a check for this :(

nomnombtc commented:
here is a something which is probably better.
It uses a Makefile target instead of a shellscript to run help2man and a few other small improvements.

https://github.com/bitcoin/bitcoin/compare/master...nomnombtc:man_automake4

>Changes compared to this here (8608):
>* Shellscript replaced by Makefile target "update-man" - many thanks to @theuni
>* DISTCHECK_CONFIGURE_FLAGS has been removed from Makefile.am, it seems not needed.
> * AC_ARG_ENABLE and AM_CONDITIONAL  in configure.ac was modified to match the look of other options.
> * The Makefile in doc/man now only installs bitcoin-qt.1  if ENABLE_QT is set.

Please tell me if you guys like this one better :) If this is ok then only the text for release-process.md is missing. I have something [here](https://github.com/nomnombtc/bitcoin/blob/man_rptmp/doc/release-process.md) but it is not ready yet, the formatting is weird of the whole section.

laanwj commented:
Thanks for sticking with this, ACK d19583f
